### ISSUE TITLE
Using NBI with TeamCity - "DLL not found" error message

### ISSUE THREAD

**Issue by DrJohnT** [Seddryck/NBi#107]
When I attempt to use configure an NBi test suite in TeamCity 9.1 using the NUnit 2.6.1 runner, I get a "DLL not found" error message.  Clearly TeamCity's NUnit runner is expecting a compiled test suite like those generated using Visual Studio (i.e. a DLL).  

The only way I found I could get my NBi test suite to work with TeamCity was to use the Unique NBITs file mentioned on this page:  http://www.nbi.io/docs/run-alterative/#unique-nbits-file
Unfortunately this means I cannot run more than one test suite.

Any ideas how I should configure NBi to work with TeamCity?

**Comment by Seddryck** [comment]
I've never tried this. 

Based on the documentation of TeamCity, it's not possible to select a .nunit file so I only see two options:
1/ Create one directory for each test-suite and copy a dll named NBi.NUnit.Runtime.dll and a test-suite in each of them.
2/ Copy/paste the dll NBi.NUnit.Runtime.dll and change their names to NBi.NUnit.Runtime.1.dll, NBi.NUnit.Runtime.2.dll and provide test-suites named NBi.NUnit.Runtime.1.nbits, NBi.NUnit.Runtime.2.nbits (Not sure this way to load test-suite is still supported by NBi).

**Comment by DrJohnT** [comment]
Of course NBi.NUnit.Runtime.1.dll will have to be in the same directory as all the other framework DLLs required by NBi.   So Option 2 works (i.e. having NBi.NUnit.Runtime.1.dll and a config file called NBi.NUnit.Runtime.1.dll.config which points to your test suite).  However, the problem is that all the test suites (or at least their config files) end up in the same directory, which is a little unmanageable.  
An alternative approach would be appreciated as ours is a big BI project.

**Comment by DrJohnT** [comment]
Actually Option 2 does not work as the two test suites defined in TeamCity as:
NBi_Tests\PostDeployTests\NBi.NUnit.Runtime.ACV_CDDS.dll
NBi_Tests\PostDeployTests\NBi.NUnit.Runtime.Mirror.dll
with appropriate matching config files and NBits files, actually calls the first set of tests twice.  Clearly the interaction between the nunit runner and the NBi.NUnit.Runtime.dll does not work correctly.  Any chance of fixing this?

**Comment by Seddryck** [comment]
It's a clear limitation of TeamCity to not be able to manage NUnit project files :-)

I'll think further about this and will come back to you.

**Comment by DrJohnT** [comment]
If the NBi.NUnit.Runtime.config file could list multiple test suites to be run, then that would solve my problem.

e.g. 

``` xml
<?xml version="1.0" encoding="utf-8" ?>
<configuration>
 <configSections>
  <section name="nbi" type="NBi.NUnit.Runtime.NBiSection, NBi.NUnit.Runtime"/>
 </configSections>
 <nbi testSuite=".\PostDeployTests.ABC.nbits"/>
 <nbi testSuite=".\PostDeployTests.XYZ.nbits"/>
 <nbi testSuite=".\PostDeployTests.NFJ.nbits"/>
</configuration>
```

**Comment by Seddryck** [comment]
It will probably be a mess to run them independantly. 

Why do you need more than one test-suite?

**Comment by FuegoArtificial** [comment]
Maybe it's not what you look for, however just as an idea:
We have built a framework around NBi (with the help of ETL, SSIS) which generates the paths, config- and nunit-files for each testsuite automatically. 
References: http://www.nbi.io/docs/run-with-nunit/ and http://www.nbi.io/docs/installation-config/
The testsuites are generated via a ETL-generated genbi-script. The test-definitions reside in a DB.
If you would setup an automatic process to configure a path, config-file etc. for each testsuite (as a DLL or whatever) you might be able to have many testsuites at once, too?

**Comment by Seddryck** [comment]
@jtnn60 I've checked your idea to reference several test-suites in one config-file and I'll not follow this path. NUnit and NBi are clearly designed to handle one test-suite by one test-suite.

Back to the root:
1. why don't you add a post-build in your script to copy/paste the dlls, config and nbits file in different directories?
2. why do you need to create more than one test-suite?

**Comment by DrJohnT** [comment]
Seddryck,

I need to have multiple test suites as the data warehouse has multiple databases and cubes and we are running the tests automatically using TeamCity after a successful build/deploy/load sequence.

The good news is that I managed to get around the problem by using Powershell as the wrapper around NUnit / NBi.  Normally, TeamCity would not understand the output of NUnit when it is called in this way.  However, TeamCity do provide a optional Addin for NUnit which causes the NUnit console runner to output the test results in a manner that TeamCity can interpret correctly. 

Basically my Powershell / psake script calls each test suite in turn using the NUnit console runner and the TeamCity Addin for NUnit allows TeamCity to format the test results properly in the web GUI.

TeamCity document the Addin here.  However, they are not so clear about what you need to do.  Basically you simply copy the JetBrains.TeamCity.NUnitAddin-NUnit-2.6.4.dll to the C:\Program Files (x86)\NUnit 2.6.4\bin\addins folder.  See
https://confluence.jetbrains.com/display/TCD9/TeamCity+Addin+for+NUnit

Regards,

John
